pkg/api/customization/cluster/actions_rketemplate.go

```diff
@@ -0,0 +1,159 @@
+	"github.com/rancher/rancher/pkg/ref"
+	v3 "github.com/rancher/types/apis/management.cattle.io/v3"
+	mgmtclient "github.com/rancher/types/client/management/v3"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func (a ActionHandler) saveAsTemplate(actionName string, action *types.Action, apiContext *types.APIContext) error {
+
+	var clusterForAccessCheck mgmtclient.Cluster
+	var err error
+	if err := access.ByID(apiContext, apiContext.Version, apiContext.Type, apiContext.ID, &clusterForAccessCheck); err != nil {
+		return httperror.NewAPIError(httperror.NotFound,
+			fmt.Sprintf("failed to get cluster by id %v", apiContext.ID))
+	}
+
+	cluster, err := a.ClusterClient.Get(apiContext.ID, v1.GetOptions{})
+	if err != nil {
+		return httperror.WrapAPIError(err, httperror.NotFound,
+			fmt.Sprintf("cluster with id %v doesn't exist", apiContext.ID))
+	}
+
+	if cluster.DeletionTimestamp != nil {
+		return httperror.NewAPIError(httperror.InvalidType,
+			fmt.Sprintf("cluster with id %v is being deleted", apiContext.ID))
+	}
+	if !v3.ClusterConditionReady.IsTrue(cluster) {
+		return httperror.WrapAPIError(err, httperror.ClusterUnavailable,
+			fmt.Sprintf("cluster not ready"))
+	}
+
+	//Copy the cluster rke spec to a template and revision object
+
+	actionInput, err := parse.ReadBody(apiContext.Request)
+	if err != nil {
+		return err
+	}
+	clusterTempName, ok := actionInput["clusterTemplateName"].(string)
+	if !ok || len(clusterTempName) == 0 {
+		return httperror.NewAPIError(httperror.InvalidBodyContent, "must specify a name for the RKE Template")
+	}
+	clusterTempRevName, ok := actionInput["clusterTemplateRevisionName"].(string)
+	if !ok || len(clusterTempRevName) == 0 {
+		return httperror.NewAPIError(httperror.InvalidBodyContent, "must specify a name for the RKE Template Revision")
+	}
+
+	creatorID := apiContext.Request.Header.Get("Impersonate-User")
+
+	newTemplateObj := newClusterTemplate(clusterTempName, cluster, creatorID)
+	clusterTemplate, err := a.ClusterTemplateClient.Create(newTemplateObj)
+	if err != nil {
+		return httperror.WrapAPIError(err, httperror.ServerError,
+			fmt.Sprintf("failed to create RKE Template object"))
+	}
+
+	newTemplateRevObj := newClusterTemplateRevision(clusterTempRevName, cluster, clusterTemplate, creatorID)
+	clusterTemplateRevision, err := a.ClusterTemplateRevisionClient.Create(newTemplateRevObj)
+	if err != nil {
+		//delete the clusterTemplate created above
+		deleteErr := a.ClusterTemplateClient.Delete(clusterTemplate.Name, &metav1.DeleteOptions{})
+		if deleteErr != nil {
+			logrus.Errorf("Failed to delete the RKE Template %v", clusterTemplate.Name)
+		}
+
+		return httperror.WrapAPIError(err, httperror.ServerError,
+			fmt.Sprintf("failed to create RKE Template Revision object"))
+	}
+
+	updatedCluster := cluster.DeepCopy()
+	updatedCluster.Spec.ClusterTemplateRevisionName = ref.Ref(clusterTemplateRevision)
```

Comment:
Nit: do you think we can add it to the beginning of the loop? lines 84-86 are same as 100-102. So it can be something:

for number of retries:
 fetch cluster 
 update
 if error  - retry